File: shared/android/app/src/main/java/io/keybase/ossifrage/ChatBroadcastReceiver.java
package io.keybase.ossifrage;

import android.app.PendingIntent;
import android.app.RemoteInput;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.RequiresApi;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import io.keybase.ossifrage.modules.NativeLogger;
import keybase.Keybase;

public class ChatBroadcastReceiver extends BroadcastReceiver {
  public static String KEY_TEXT_REPLY = "key_text_reply";

  @RequiresApi(api = Build.VERSION_CODES.KITKAT_WATCH)
  private String getMessageText(Intent intent) {
    Bundle remoteInput = RemoteInput.getResultsFromIntent(intent);
    if (remoteInput != null) {
        return remoteInput.getCharSequence(KEY_TEXT_REPLY).toString();
    }
    return null;
 }

  @RequiresApi(api = Build.VERSION_CODES.KITKAT_WATCH)
  @Override
  public void onReceive(Context context, Intent intent) {
    MainActivity.setupKBRuntime(context, false);
    ConvData convData = new ConvData(intent);
    PendingIntent openConv = intent.getParcelableExtra("openConvPendingIntent");
    NotificationCompat.Builder repliedNotification = new NotificationCompat.Builder(context, KeybasePushNotificationListenerService.CHAT_CHANNEL_ID)
      .setContentIntent(openConv)
      .setTimeoutAfter(1000)
      .setSmallIcon(R.drawable.ic_notif);
    NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);

    String messageBody = getMessageText(intent);
    if (messageBody != null) {

      try {
        WithBackgroundActive withBackgroundActive = () -> Keybase.handlePostTextReply(convData.convID, convData.tlfName, convData.lastMsgId, messageBody);
        withBackgroundActive.whileActive(context);
        repliedNotification.setContentText("Replied");
      } catch (Exception e) {
        repliedNotification.setContentText("Couldn't send reply");
        NativeLogger.error("Failed to send quick reply", e);
      }

    } else {
      repliedNotification.setContentText("Couldn't send reply - Failed to read input.");
      NativeLogger.error("Message Body in quick reply was null");
    }

    notificationManager.notify(convData.convID, 0, repliedNotification.build());
  }
}

class ConvData {
  String convID;
  String tlfName;
  long lastMsgId;

  ConvData(String convId, String tlfName, long lastMsgId) {
    this.convID = convId;
    this.tlfName = tlfName;
    this.lastMsgId = lastMsgId;
  }

  ConvData (Intent intent) {
    Bundle data = intent.getBundleExtra("ConvData");
    this.convID = data.getString("convID");
    this.tlfName = data.getString("tlfName");
    this.lastMsgId = data.getLong("lastMsgId");
  }

  public Intent intoIntent(Context context) {
    Bundle data = new Bundle();
    data.putString("convID", this.convID);
    data.putString("tlfName", this.tlfName);
    data.putLong("lastMsgId", this.lastMsgId);
    Intent intent = new Intent(context, ChatBroadcastReceiver.class);
    intent.putExtra("ConvData", data);
    return intent;
  }
}

File: shared/android/app/src/main/java/io/keybase/ossifrage/ChatBroadcastReceiver.kt
package io.keybase.ossifrage

import android.app.PendingIntent
import android.app.RemoteInput
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import io.keybase.ossifrage.MainActivity.Companion.setupKBRuntime
import io.keybase.ossifrage.modules.NativeLogger
import keybase.Keybase

class ChatBroadcastReceiver : BroadcastReceiver() {
    private fun getMessageText(intent: Intent): String? {
        val remoteInput = RemoteInput.getResultsFromIntent(intent)
        return remoteInput?.getCharSequence(KEY_TEXT_REPLY)?.toString()
    }

    override fun onReceive(context: Context, intent: Intent) {
        val convData = ConvData.fromIntent(intent)
        val openConv = intent.getParcelableExtra<PendingIntent>("openConvPendingIntent")
        val messageBody = getMessageText(intent)
        val pendingResult = goAsync()
        runReceiverWork(RECEIVER_BUDGET_MS, { Thread(it).start() }, { NativeLogger.warn(it) }, { msg, e -> NativeLogger.error(msg, e) },
                { pendingResult.finish() }) {
            val status = if (messageBody == null) {
                NativeLogger.error("Message Body in quick reply was null")
                "Couldn't send reply - Failed to read input."
            } else {
                setupKBRuntime(context, false)
                sendQuickReply({ msg, e -> NativeLogger.error(msg, e) }) {
                    postTextReplyInPushWindow(context, convData, messageBody)
                }
            }
            val repliedNotification = NotificationCompat.Builder(context, KeybasePushNotificationListenerService.CHAT_CHANNEL_ID)
                    .setContentIntent(openConv)
                    .setTimeoutAfter(1000)
                    .setSmallIcon(R.drawable.ic_notif)
                    .setContentText(status)
            NotificationManagerCompat.from(context).notify(convData.convID, 0, repliedNotification.build())
        }
    }

    // Transitional: the push window is opened here, for the same reason as
    // WithBackgroundActive -- it goes away once the bind layer wraps the reply in the
    // window itself. Unlike WithBackgroundActive this never skips the send while the app
    // is foreground; a reply typed in the notification shade must go out either way.
    private fun postTextReplyInPushWindow(context: Context, convData: ConvData, messageBody: String) {
        // 0 when the app is active and nothing needs holding up.
        val token = Keybase.appPushWindowBegin()
        try {
            Keybase.handlePostTextReply(convData.convID, convData.tlfName, convData.lastMsgId, messageBody)
        } finally {
            if (token > 0) {
                // Hands over to a background task if the UI is still in the background and
                // work must keep going.
                Keybase.appPushWindowEnd(token, KBPushNotifier(context, Bundle()))
            }
        }
    }

    companion object {
        const val KEY_TEXT_REPLY = "key_text_reply"

        // goAsync gives a broadcast 10s; leave margin.
        private const val RECEIVER_BUDGET_MS = 9_000L
    }
}

internal data class ConvData(
    @JvmField val convID: String?,
    val tlfName: String?,
    val lastMsgId: Long
) {
    fun intoIntent(context: Context?): Intent {
        val data = Bundle()
        data.putString("convID", convID)
        data.putString("tlfName", tlfName)
        data.putLong("lastMsgId", lastMsgId)
        val intent = Intent(context, ChatBroadcastReceiver::class.java)
        intent.putExtra("ConvData", data)
        return intent
    }

    companion object {
        fun fromIntent(intent: Intent): ConvData {
            val data = intent.getBundleExtra("ConvData")!!
            return ConvData(
                convID = data.getString("convID"),
                tlfName = data.getString("tlfName"),
                lastMsgId = data.getLong("lastMsgId")
            )
        }
    }
}
